Frost-text test suite

Jasonjys opened the pull request:
created integration and unit tests
updated the README
cleaned up frost-text
updated the dummy app

Jasonjys committed:
added default properties tests

Jasonjys committed:
added intergration tests for disabled, error and placeholder

Jasonjys committed:
added frost-text test and updated README

Jasonjys committed:
fixed focusin test

Jasonjys committed:
used run loop to fix focusIn test

Jasonjys committed:
removed debugger

Jasonjys committed:
removed hook

Jasonjys committed:
removed comments

Jasonjys committed:
fixed some code

Jasonjys committed:
fixed indentation of integration test

Jasonjys committed:
changed API to alphabetical order

Jasonjys committed:
added example of "readonly" under State section and "autofocus" under Options section in demo

Jasonjys committed:
added example of "value" under Options section in demo

Jasonjys committed:
added "readonly" to API

Jasonjys committed:
1. user can no longer clear when "readonly" attribute is set to true
2. updated unit and integration tests for "readonly" property

Jasonjys committed:
fixed syntax error and added tests for "autocapitalize" and "autocorrect" etc...

Jasonjys committed:
added "spellcheck" attribute to API

Jasonjys committed:
fixed "spellcheck" in unit test

Jasonjys committed:
fixed syntax errors

Jasonjys committed:
Merge remote-tracking branch 'Michael/LTS-prep' into textTest

Jasonjys committed:
Merge remote-tracking branch 'Michael/LTS-prep' into textTest

Jasonjys committed:
fixed error in README

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+    })
+
+    it('sets align property', function () {
+      const align = 'right'
```
With the const up here I would expect that to be used in the hasClass below so that the test and the variable reflect each other

Jasonjys replied:
fixed

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+    })
+
+    it('sets autofocus property', function () {
+      const autofocus = 'true'
```
Is string `'true'` the same as boolean `true`?  I'd probably use the boolean here

Jasonjys replied:
You are right. Fixed.

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+    it('set disabled property', function () {
+      this.render(hbs`
+        {{frost-text
+          disabled=true
```
No `const` in this case?

Jasonjys replied:
The convention Michael and I are following now is if the property is a boolean we set it directly and if it is a string we set it using the controller. :)

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+      expect(
+        this.$('input').attr('readonly'),
+        'readonly attribute is set'
+      ).to.eql('readonly')
```
Does this pass?  I would have expected it to either equal `true` (from above) or null (if the value is stripped)

Jasonjys replied:
Yes, it pass. In this case, setting "readonly" only set the word which is a little bit special. So if I set readonly=true, only "readonly" shows up but not "readonly=true".

juwara0 replied:
Once you change the `.attr()` to `.prop()` you can change the expect `.to.be.true`

Jasonjys replied:
fixed

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+      expect(
+        this.$('input').attr('required'),
+        'required attribute is set'
+      ).to.eql('required')
```
Same comment as `readonly`

Jasonjys replied:
Same thing as readonly.

juwara0 replied:
Once you change the .attr() to .prop() you can change the expect .to.be.true

Jasonjys replied:
fixed

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+    })
+
+    it('set spellcheck property', function () {
+      const spellcheck = 'true'
```
boolean

Jasonjys replied:
We changed it to boolean. It is a bit confusing because the spellcheck attribute is an enumerated attribute whose keywords are the empty string, true and false. 
https://html.spec.whatwg.org/#spelling-and-grammar-checking

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+      `)
+
+      expect(
+        this.$('.frost-text').hasClass('error'),
```
reference the const

Jasonjys replied:
We removed this test because we don't need to test ember.

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+      this.render(hbs`
+        {{frost-text}}
+      `)
+
```
We should check that it's **not** present prior to the input trigger

Jasonjys replied:
Fixed

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+      ).to.be.true
+    })
+
+    it('text cleared on button click', function () {
```
The test description doesn't match the test body

Jasonjys replied:
fixed

sglanzer-fire commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+
+      this.$('input').trigger('input')
+
+      const testObj = {
```
Why create the testObj if you're going to just refer to the individual properties?

juwara0 replied:
fixed

sglanzer-fire reviewed (changes requested):
Take a look at the comments and do a cleanup pass over the tests and we'll re-review

sglanzer-fire commented:
Good first effort, there is some cleanup needed, but it's good to see you produce a test suite like this :)

Jasonjys commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,318 @@ describeComponent(
+    })
+
+    it('sets autofocus property', function () {
+      const autofocus = 'true'
```
fixed

Jasonjys committed:
fixes asked as in code review

Jasonjys committed:
Merge remote-tracking branch 'Michael/LTS-prep' into textTest

Jasonjys committed:
fixed ordering

juwara0 commented on docs/frost-text.md:
```diff
@@ -7,22 +7,33 @@
 | Attribute | Type | Value | Description |
 | --------- | ---- | ----- | ----------- |
-| `value` | `string` | `<value-text>` | text to be displayed in text field |
 | `align` |`string` | `right` | right align text input |
```
There are two values for `align`. The default is `left` the other value is `right`. We should list both.

juwara0 replied:
fixed

juwara0 commented on docs/frost-text.md:
```diff
@@ -7,22 +7,33 @@
-| `placeholder` | `string` | `<text>` | placeholder text |
+| `autocapitalize` |`string` | `off` | **default** - normal text field |
+ | | `on` | text will be automatically capitalized (supported by Chrome and iOS Safari Mobile) |
+ | `autocorrect` |`string` | `off` | **default** - normal text field |
```
The indentation is off on this. We need to remove the extra space at the beginning.

juwara0 replied:
the same problem exists for the line above this one and the line below this one.

juwara0 replied:
fixed

juwara0 commented on docs/frost-text.md:
```diff
@@ -7,22 +7,33 @@
-| `onFocus` | `string` | `<action-name>` | triggers associated action when component receives focus |
-| `onKeyDown` | `string` | `<action-name>` | triggers associated action when component receive a `keydown` event |
-| `onKeyUp` | `string` | `<action-name>` | triggers associated action when component receive a `keyup` event |
+| `maxlength` | `string` | `<maxlength-value>` | maximum number of characters a user can enter |
```
Lets change this to a `number`.

Jasonjys replied:
fixed

juwara0 commented on docs/frost-text.md:
```diff
@@ -7,22 +7,33 @@
+| `spellcheck` | `string` | `false` | **default** - normal text field |
+| | | `true` | spelling and grammar are checked |
+| | | `default` | defer to default behavior |
+| `tabindex` | `string` | `<tabindex-value>` | the tabindex value |
```
Lets change this to a number.

Jasonjys replied:
fixed

juwara0 commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,301 @@ describeComponent(
+      `)
+
+      expect(
+        this.$('input').attr('tabindex'),
```
Lets change this to use the `.prop()` method instead of `.attr()` then once the prop-types is set to number and the default is set to `0` you can change the expect to `.to.eql(0)`

juwara0 replied:
fixed

juwara0 commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,301 @@ describeComponent(
+      ).to.eql('0')
+
+      expect(
+        this.$('input').attr('type'),
```
Change all cases of `.attr()` to `.prop()`

Jasonjys replied:
Everything worked except "autocorrect" after I changed .attr() to .prop(). 
I am getting this: AssertionError: autocorrect attribute is set: expected undefined to deeply equal 'on'
Did some research and it says it needs apple extension in order to work in the Safari HTML Reference. 
Please see the autocorrect section: 
https://developer.apple.com/library/content/documentation/AppleApplications/Reference/SafariHTMLRef/Articles/Attributes.html#//apple_ref/doc/uid/TP40008058-autocorrect

juwara0 commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,301 @@ describeComponent(
+      `)
+
+      expect(
+        this.$('input').attr('autocapitalize'),
```
same comment about changing the `.attr()` to `.prop()`. Please make that change for every occurrence.

Jasonjys replied:
It doesn't work when I changed it to .prop().
AssertionError: autocapitalize attribute is set: expected 'sentences' to deeply equal 'on'
I did some searching and it has something to do with the default value of "autocapitalize" on mobile.
The default and invalid value for "autocapitalize" is "sentences" and it can be set to "off", "characters" and "words".
https://developers.google.com/web/updates/2015/04/autocapitalize?hl=en

juwara0 commented on tests/integration/components/frost-text-test.js:
```diff
@@ -16,29 +19,301 @@ describeComponent(
+    })
+
+    it('sets maxlength property', function () {
+      const maxlength = '30'
```
Please change this to remove the quotes:
`const maxlength = 30`

Jasonjys replied:
fixed

juwara0 commented on tests/unit/components/frost-text-test.js:
```diff
@@ -23,5 +32,124 @@ describeComponent(
+      expect(
+        component.get('tabindex'),
+        'tabindex: 0'
+      ).to.be.eql('0')
```
You can change this to `0` once you update the default value to not be a string.

Jasonjys replied:
fixed

Jasonjys reviewed (commented):
1. changed "maxlength", "tabindex" to number
2. updated "align" property and fixed indentation in API
3. changed .attr() to .prop() ("autocorrect" is not working)

Jasonjys committed:
1. changed "maxlength", "tabindex" to number
2. updated "align" property and fixed indentation in API
3. changed .attr() to .prop() ("autocorrect" is not working)

Jasonjys committed:
1. commented out "autocapitalize" and "autocorrect" integration tests
2. updated API

Jasonjys committed:
1. added integration test for "tabindex"
2. updated dummy app
3. changed .attr() to .prop()

Jasonjys committed:
1. fixed integration test for "tabindex"
2. added integration test for "title"
3. updated test for "calls onInput closure action"

Jasonjys committed:
clean up

juwara0 committed:
Merge remote-tracking branch 'origin/LTS-prep' into textTest
